# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студенты гр. 2383 Борисов Иван Вадимович и Сыздыков Нургалым Конакбаевич. Вариант №25

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Краткое изложение основных теоретических понятий

В статистике выделяют несколько типов связи между случайными величинами $\{X, Y\}$:

- Функциональная зависимость: Каждому значению одной величины соответствует строго определенное значение другой ($y=g(x)$).
- Статистическая зависимость: Изменение одной величины влечет за собой изменение закона распределения другой.
- Корреляционная зависимость: Частный случай статистической связи, при котором изменение одной величины приводит к изменению математического ожидания другой. Функции, описывающие эту связь, называются функциями регрессии.

Для оценки тесноты линейной связи используется коэффициент корреляции $r_{xy}$:
- Значение коэффициента находится в диапазоне $[-1,1]$.
- Если $∣r_{xy}∣ = 1$, связь является функциональной.
- Если $r_{xy} > 0$, связь прямая, если $r_{xy} < 0$ $-$ обратная.

Методы расчета выборочного значения $r_{xy}$:
Стандартная формула:
$$
\overline{r}_{xy} = \frac{\sum_{i=1}^{K_y}\sum_{j=1}^{K_x} n_{ij} y_i x_j - N \overline{x}_в \overline{y}_в} {N S_x S_y}
$$

Метод условных вариант:
$$
\overline{r}_{xy} = \frac{\sum_{i=1}^{K_y}\sum_{j=1}^{K_x} n_{ij} u_i v_j - N \overline{u}_в \overline{v}_в} {N \overline{\sigma}_u \overline{\sigma}_v}
$$

где
- $u_i = \frac{x_i - C_x}{h_x}$;
- $v_j = \frac{y_j - C_y}{h_y}$.

условные варианты.

При сильной корреляции распределение $r_{xy}$ асимметрично, поэтому используется z-преобразование Фишера для перехода к нормально распределенной величине:
- Прямое преобразование:
$$
\overline{z} = 0.5 \ln {\frac{1 + \overline{r}_{xy}}{1 - \overline{r}_{xy}}}
$$
- СКО величины z:
$$
\overline{\sigma}_z = \frac{1}{\sqrt{N−3}}
$$
- Доверительный интервал для z:
$$
(\overline{z} − \lambda(\gamma) \overline{\sigma}_z, \overline{z} + \lambda(\gamma) \overline{\sigma}_z)
$$
- Обратное преобразование:
$$
r = \tanh(z) = \frac{e^{2z}-1}{e^{2z}+1}
$$
  
Статистическая гипотеза — утверждение о виде распределения или параметрах генеральной совокупности.
- Основная гипотеза ($H_0$): $r_{xy} = 0$
- Альтернативная гипотеза ($H_1$): $r_{xy} \neq 0$

Для проверки используется критерий Стьюдента:
$$
T = \frac{\overline{r}_{xy} \sqrt{N - 2}}{\sqrt{1 - {\overline{r}_{xy}}^2}}
$$

Если $∣T_{набл}∣ > t_{крит}(\alpha,k)$, то нулевая гипотеза отвергается, и корреляция признается значимой.

## Постановка задачи с кратким описанием порядка выполнения работы

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин X (первый признак) и Y (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

### Порядок выполнения работы

1. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации). Оформить результаты в виде таблицы, сделать выводы.
2. Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.
3. По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.
4. Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.
5. Построить доверительный интервал для коэффициента корреляции при уровне значимости $\gamma \in \{0.95,0.99\}$, сделать выводы.
6. Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости $\alpha = 0.05$, сделать выводы.

### 1. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации). Оформить результаты в виде таблицы, сделать выводы.

In [1]:
import numpy as np
import pandas as pd
from scipy import stats
from IPython.display import display, Markdown

pd.options.display.float_format = '{:,.4f}'.format

In [2]:
df: pd.DataFrame = pd.read_csv('Syzdykov_Borisov_data_116.csv')

In [3]:
df

,nu,E
0,396,90.1000
1,525,145.3000
2,437,129.4000
3,411,115.2000
4,510,124.1000
...,...,...
111,463,144.9000
112,487,146.0000
113,528,163.4000
114,331,74.1000


In [4]:
def get_interval_stats(series: pd.Series) -> tuple[np.ndarray, np.ndarray, float, float]:
    n: int = len(series)
    k: int = int(np.ceil(1 + 3.31 * np.log10(n)))

    x_min: float = series.min()
    x_max: float = series.max()
    h: float = (x_max - x_min) / k

    bins: np.ndarray = np.linspace(x_min, x_max, k + 1)
    counts, edges = np.histogram(series, bins=bins)
    midpoints: np.ndarray = (edges[:-1] + edges[1:]) / 2

    return midpoints, counts, h, x_min

In [5]:
def create_table_1(midpoints: np.ndarray, counts: np.ndarray, h: float, label: str) -> tuple[pd.DataFrame, float, float]:
    """Формирование расчетной таблицы №1 с проверкой контроля."""
    max_idx: int = np.argmax(counts)
    C: float = midpoints[max_idx]

    # Расчет условных вариант
    u: np.ndarray = np.round((midpoints - C) / h).astype(int)

    n_u: np.ndarray = counts * u
    n_u2: np.ndarray = counts * (u**2)
    n_u3: np.ndarray = counts * (u**3)
    n_u4: np.ndarray = counts * (u**4)
    n_u_plus_1_4: np.ndarray = counts * ((u + 1)**4)

    df_t1: pd.DataFrame = pd.DataFrame({
        r'$i$': range(1, len(midpoints) + 1),
        r'$x_i$': midpoints,
        r'$n_i$': counts,
        r'$u_i$': u,
        r'$n_i \cdot u_i$': n_u,
        r'$n_i \cdot u_i^2$': n_u2,
        r'$n_i \cdot u_i^3$': n_u3,
        r'$n_i \cdot u_i^4$': n_u4,
        r'$n_i \cdot (u_i+1)^4$': n_u_plus_1_4
    })

    sums: dict = df_t1.sum()
    sum_row: pd.DataFrame = pd.DataFrame([{
        '$i$': r'$\Sigma$', r'$x_i$': np.nan, r'$n_i$': sums[r'$n_i$'], r'$u_i$': np.nan,
         r'$n_i \cdot u_i$': sums[r'$n_i \cdot u_i$'], r'$n_i \cdot u_i^2$': sums[r'$n_i \cdot u_i^2$'],
         r'$n_i \cdot u_i^3$': sums[r'$n_i \cdot u_i^3$'], r'$n_i \cdot u_i^4$': sums[r'$n_i \cdot u_i^4$'],
         r'$n_i \cdot (u_i+1)^4$': sums[r'$n_i \cdot (u_i+1)^4$']
    }])

    lhs: float = sums[r'$n_i \cdot (u_i+1)^4$']
    rhs: float = (sums[r'$n_i \cdot u_i^4$'] + 4*sums[r'$n_i \cdot u_i^3$'] + 6*sums[r'$n_i \cdot u_i^2$'] + 4*sums[r'$n_i \cdot u_i$'] + sums[r'$n_i$'])

    print(f"--- Контроль вычислений для {label} ---")
    display(Markdown(rf"$\sum n_i (u_i+1)^4 = {lhs}$"))
    display(Markdown(rf"$\sum n_i \cdot u_i^4 + 4 \sum n_i \cdot u_i^3 + 6 \sum n_i \cdot u_i^2 + 4 \sum n_i \cdot u_i + N = {rhs}$"))
    print(rf"Контроль пройден: {np.isclose(lhs, rhs)}")

    return pd.concat([df_t1, sum_row], ignore_index=True), C, h

In [6]:
def calculate_moments(table1: pd.DataFrame, h: float) -> pd.DataFrame:
    """Вычисление условных и центральных моментов."""
    row_sum: pd.Series = table1.iloc[-1]
    n: float = row_sum[r'$n_i$']

    # Условные моменты
    M1_star: float = row_sum[r'$n_i \cdot u_i$'] / n
    M2_star: float = row_sum[r'$n_i \cdot u_i^2$'] / n
    M3_star: float = row_sum[r'$n_i \cdot u_i^3$'] / n
    M4_star: float = row_sum[r'$n_i \cdot u_i^4$'] / n

    # Центральные моменты (через условные)
    m1: float = 0.0
    m2: float = (M2_star - M1_star**2) * (h**2)
    m3: float = (M3_star - 3*M1_star*M2_star + 2*M1_star**3) * (h**3)
    m4: float = (M4_star - 4*M1_star*M3_star + 6*M1_star**2*M2_star - 3*M1_star**4) * (h**4)

    return pd.DataFrame({
        r'$k$': [1, 2, 3, 4],
        r'$\overline{M}^*_k$': [M1_star, M2_star, M3_star, M4_star],
        r'$\overline{m}_k$': [m1, m2, m3, m4]
    })

In [7]:
def compute_main_stats(midpoints: np.ndarray, counts: np.ndarray, moments: pd.DataFrame, C: float, h: float, label: str) -> tuple[float, float, float]:
    """Сравнение методов расчета внутри интервального ряда."""
    # 1. Метод условных вариант
    m1_star: float = moments.loc[0, r'$\overline{M}^*_k$']
    x_avg_cond: float = m1_star * h + C
    d_v_cond: float = moments.loc[1, r'$\overline{m}_k$']

    # 2. Стандартная формула для интервального ряда
    x_avg_int: float = np.average(midpoints, weights=counts)
    d_v_int: float = np.average((midpoints - x_avg_int)**2, weights=counts)

    # 3. Выборочное СКО
    sigma_v: float = np.sqrt(d_v_cond)

    display(Markdown(f"#### Результаты для {label}:"))

    comparison_df: pd.DataFrame = pd.DataFrame({
        "Показатель": ["Среднее значение", "Выборочная дисперсия"],
        "Стандартная формула": [x_avg_int, d_v_int],
        "Условные варианты": [x_avg_cond, d_v_cond],
        "Разность": [abs(x_avg_int - x_avg_cond), abs(d_v_int - d_v_cond)]
    })
    display(comparison_df)

    display(Markdown(rf"Выборочное СКО ($\sigma_в$): **{sigma_v:.4f}**"))
    return x_avg_cond, d_v_cond, sigma_v

In [8]:
def calculate_corrected_estimates(d_v: float, n: int, label: str) -> tuple[float, float]:
    s2: float = (n / (n - 1)) * d_v
    s: float = np.sqrt(s2)

    display(Markdown(f"#### Исправленные оценки для {label}:"))
    display(Markdown(rf"- Исправленная дисперсия ($s^2$): **{s2:.4f}**"))
    display(Markdown(rf"- Исправленное СКО ($s$): **{s:.4f}**"))

    return s2, s

In [9]:
def analyze_distribution_shape(moments: pd.DataFrame, sigma_v: float, label: str) -> None:
    """Расчет асимметрии и эксцесса."""
    m3: float = moments.loc[2, r'$\overline{m}_k$']
    m4: float = moments.loc[3, r'$\overline{m}_k$']

    # Асимметрия и эксцесс (используем выборочное СКО)
    as_val: float = m3 / (sigma_v**3)
    e_val: float = (m4 / (sigma_v**4)) - 3

    display(Markdown(f"#### Характеристики формы для {label}:"))
    display(Markdown(rf"- Коэффициент асимметрии ($\overline{{A}}_s$): **{as_val:.4f}**"))
    display(Markdown(rf"- Коэффициент эксцесса ($\overline{{E}}$): **{e_val:.4f}**"))

In [10]:
def compute_robust_and_variation(
    midpoints: np.ndarray,
    counts: np.ndarray,
    h: float,
    x_avg: float,
    sigma_v: float,
    label: str
) -> None:
    """Расчет моды, медианы и коэффициента вариации."""
    # 1. Коэффициент вариации
    v_star: float = (sigma_v / x_avg) * 100

    # 2. Мода (Mo)
    max_idx: int = np.argmax(counts)
    # Нижняя граница модального интервала
    l_mo: float = midpoints[max_idx] - h/2
    n_mo: int = counts[max_idx]
    n_mo_prev: int = counts[max_idx - 1] if max_idx > 0 else 0
    n_mo_next: int = counts[max_idx + 1] if max_idx < len(counts) - 1 else 0

    mo_star: float = l_mo + h * (n_mo - n_mo_prev) / ((n_mo - n_mo_prev) + (n_mo - n_mo_next))

    # 3. Медиана (Me)
    half_n: float = np.sum(counts) / 2
    cum_counts: np.ndarray = np.cumsum(counts)

    me_idx: int = np.where(cum_counts >= half_n)[0][0]
    l_me: float = midpoints[me_idx] - h/2
    s_me_prev: float = cum_counts[me_idx - 1] if me_idx > 0 else 0
    n_me: int = counts[me_idx]

    me_star: float = l_me + h * (half_n - s_me_prev) / n_me

    display(Markdown(f"#### Итоговые показатели для {label}:"))
    display(Markdown(rf"- Мода ($M_o^*$): **{mo_star:.4f}**"))
    display(Markdown(rf"- Медиана ($M_e^*$): **{me_star:.4f}**"))
    display(Markdown(rf"- Коэффициент вариации ($V^*$): **{v_star:.2f}%**"))

In [11]:
# Мат ожидание
mid_e, counts_e, h_e, _ = get_interval_stats(df['E'])

# Таблица признака
table1_e, C_e, h_e_final = create_table_1(mid_e, counts_e, h_e, "признака E")
display(Markdown("**Таблица 1 для признака E:**"))
display(table1_e)

# Моменты
moments_e = calculate_moments(table1_e, h_e_final)
display(Markdown("**Таблица 2 (Моменты) для признака E:**"))
display(moments_e)

# Результаты
x_e, d_e, sigma_e = compute_main_stats(mid_e, counts_e, moments_e, C_e, h_e_final, "E")
s2_e, s_e = calculate_corrected_estimates(d_e, len(df), "E")
analyze_distribution_shape(moments_e, sigma_e, "E")
compute_robust_and_variation(mid_e, counts_e, h_e_final, x_e, sigma_e, "E")

summary_table = pd.DataFrame({
    'Параметр': [
        r'Математическое ожидание ($\overline{x}_в$)', 
        r'Дисперсия ($D_в$)', 
        r'Исправленная дисперсия ($s^2$)',
        r'Выборочное СКО ($\sigma_в$)', 
        r'Исправленное СКО ($s$)',
        r'Коэффициент асимметрии ($\overline{A}_s$)', 
        r'Коэффициент эксцесса ($\overline{E}$)', 
        r'Мода ($M_o^*$)', 
        r'Медиана ($M_e^*$)', 
        r'Коэффициент вариации ($V^*$)'
    ],
    'Значение': [
        128.1636, 480.7445, 484.9249, 21.9259, 22.0210, 
        -0.3290, 0.0064, 133.4000, 130.1934, '17.11%'
    ]
})

display(Markdown("### Итоговая таблица точечных оценок для признака E"))
display(summary_table)

--- Контроль вычислений для признака E ---


$\sum n_i (u_i+1)^4 = 2955.0$

$\sum n_i \cdot u_i^4 + 4 \sum n_i \cdot u_i^3 + 6 \sum n_i \cdot u_i^2 + 4 \sum n_i \cdot u_i + N = 2955.0$

Контроль пройден: True


**Таблица 1 для признака E:**

,$i$,$x_i$,$n_i$,$u_i$,$n_i \cdot u_i$,$n_i \cdot u_i^2$,$n_i \cdot u_i^3$,$n_i \cdot u_i^4$,$n_i \cdot (u_i+1)^4$
0,1,78.8875,6.0000,-3.0000,-18.0000,54.0000,-162.0000,486.0000,96.0000
1,2,94.4625,10.0000,-2.0000,-20.0000,40.0000,-80.0000,160.0000,10.0000
2,3,110.0375,15.0000,-1.0000,-15.0000,15.0000,-15.0000,15.0000,0.0000
3,4,125.6125,34.0000,0.0000,0.0000,0.0000,0.0000,0.0000,34.0000
4,5,141.1875,34.0000,1.0000,34.0000,34.0000,34.0000,34.0000,544.0000
5,6,156.7625,14.0000,2.0000,28.0000,56.0000,112.0000,224.0000,"1,134.0000"
6,7,172.3375,2.0000,3.0000,6.0000,18.0000,54.0000,162.0000,512.0000
7,8,187.9125,1.0000,4.0000,4.0000,16.0000,64.0000,256.0000,625.0000
8,$\Sigma$,NaN,116.0000,NaN,19.0000,233.0000,7.0000,"1,337.0000","2,955.0000"


**Таблица 2 (Моменты) для признака E:**

,$k$,$\overline{M}^*_k$,$\overline{m}_k$
0,1,0.1638,0.0000
1,2,2.0086,480.7445
2,3,0.0603,"-3,467.8573"
3,4,11.5259,"694,816.1355"


#### Результаты для E:

,Показатель,Стандартная формула,Условные варианты,Разность
0,Среднее значение,128.1636,128.1636,0.0000
1,Выборочная дисперсия,480.7445,480.7445,0.0000


Выборочное СКО ($\sigma_в$): **21.9259**

#### Исправленные оценки для E:

- Исправленная дисперсия ($s^2$): **484.9249**

- Исправленное СКО ($s$): **22.0210**

#### Характеристики формы для E:

- Коэффициент асимметрии ($\overline{A}_s$): **-0.3290**

- Коэффициент эксцесса ($\overline{E}$): **0.0064**

#### Итоговые показатели для E:

- Мода ($M_o^*$): **133.4000**

- Медиана ($M_e^*$): **130.1934**

- Коэффициент вариации ($V^*$): **17.11%**

### Итоговая таблица точечных оценок для признака E

,Параметр,Значение
0,Математическое ожидание ($\overline{x}_в$),128.1636
1,Дисперсия ($D_в$),480.7445
2,Исправленная дисперсия ($s^2$),484.9249
3,Выборочное СКО ($\sigma_в$),21.9259
4,Исправленное СКО ($s$),22.0210
5,Коэффициент асимметрии ($\overline{A}_s$),-0.3290
6,Коэффициент эксцесса ($\overline{E}$),0.0064
7,Мода ($M_o^*$),133.4000
8,Медиана ($M_e^*$),130.1934
9,Коэффициент вариации ($V^*$),17.11%


На основе таблицы точечных оценок для признака E можно сделать следующие выводы:
- Средние показатели: Средний модуль упругости исследуемой древесины составляет $128.16 \frac{кг}{{см}^2}$.
- Однородность выборки: Коэффициент вариации $V^∗ =17.11 \%$ ниже порога в 33%. Это указывает на однородность данных и репрезентативность среднего значения, что согласуется с результатами анализа плотности $\nu$ в предыдущих работах.
- Характер асимметрии: Отрицательный коэффициент $\overline{A}_s = -0.3290$ в сочетании с неравенством $M_o^∗ > M_e^∗ > \overline{x}_в$ указывает на левостороннюю асимметрии распределения.
- Крутизна распределения: Коэффициент эксцесса $E = 0.0064$ близок к нулю, что говорит о близости формы распределения к нормальному закону.
- Смещение оценок: Исправленное СКО ($s = 22.0210$) превышает выборочное ($\sigma_в = 21.9259$), что позволяет использовать его как несмещенную оценку для дальнейшего построения доверительных интервалов.

### 2. Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.

In [12]:
def build_bivariate_series_intervals(df, col_x, col_y):
    def get_bins(series):
        n = len(series)
        k = int(np.ceil(1 + 3.31 * np.log10(n)))
        return np.linspace(series.min(), series.max(), k + 1)

    bins_x = get_bins(df[col_x])
    bins_y = get_bins(df[col_y])

    labels_x = [f"[{bins_x[i]:.2f}, {bins_x[i+1]:.2f})" for i in range(len(bins_x)-2)]
    labels_x.append(f"[{bins_x[-2]:.2f}, {bins_x[-1]:.2f}]")
    
    labels_y = [f"[{bins_y[i]:.2f}, {bins_y[i+1]:.2f})" for i in range(len(bins_y)-2)]
    labels_y.append(f"[{bins_y[-2]:.2f}, {bins_y[-1]:.2f}]")

    x_intervals = pd.cut(df[col_x], bins=bins_x, labels=labels_x, include_lowest=True)
    y_intervals = pd.cut(df[col_y], bins=bins_y, labels=labels_y, include_lowest=True)

    return x_intervals, y_intervals

col_x = 'nu'
col_y = 'E'
x_intervals, y_intervals = build_bivariate_series_intervals(df, col_x, col_y)
bivariate_table = pd.crosstab(
    x_intervals, 
    y_intervals, 
    dropna=False,
    rownames=[f'Признак {col_x}'], 
    colnames=[f'Признак {col_y}']
)

print("Двумерный интервальный вариационный ряд:")
display(bivariate_table)

Двумерный интервальный вариационный ряд:


Признак E,"[71.10, 86.67)","[86.67, 102.25)","[102.25, 117.82)","[117.82, 133.40)","[133.40, 148.97)","[148.97, 164.55)","[164.55, 180.12)","[180.12, 195.70]"
Признак nu,,,,,,,,
"[320.00, 357.88)",3,3,1,0,0,0,0,0
"[357.88, 395.75)",1,4,4,0,0,0,0,0
"[395.75, 433.62)",2,3,8,9,0,0,0,0
"[433.62, 471.50)",0,0,1,21,11,0,0,0
"[471.50, 509.38)",0,0,1,1,17,7,0,0
"[509.38, 547.25)",0,0,0,3,5,5,1,0
"[547.25, 585.12)",0,0,0,0,1,1,1,0
"[585.12, 623.00]",0,0,0,0,0,1,0,1


### 3. По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.

In [18]:
correlation_table = pd.crosstab(
    x_intervals, 
    y_intervals, 
    margins=True, 
    margins_name="n_y (сумма)",
    rownames=[f'Признак nu'], 
    colnames=[f'Признак E']
)

correlation_table.rename(columns={'$n_y$ (сумма)': '$n_x$ (сумма)'}, inplace=True)

display(Markdown("### Корреляционная таблица"))
display(correlation_table)

### Корреляционная таблица

Признак E,"[71.10, 86.67)","[86.67, 102.25)","[102.25, 117.82)","[117.82, 133.40)","[133.40, 148.97)","[148.97, 164.55)","[164.55, 180.12)","[180.12, 195.70]",n_y (сумма)
Признак nu,,,,,,,,,
"[320.00, 357.88)",3,3,1,0,0,0,0,0,7
"[357.88, 395.75)",1,4,4,0,0,0,0,0,9
"[395.75, 433.62)",2,3,8,9,0,0,0,0,22
"[433.62, 471.50)",0,0,1,21,11,0,0,0,33
"[471.50, 509.38)",0,0,1,1,17,7,0,0,26
"[509.38, 547.25)",0,0,0,3,5,5,1,0,14
"[547.25, 585.12)",0,0,0,0,1,1,1,0,3
"[585.12, 623.00]",0,0,0,0,0,1,0,1,2
n_y (сумма),6,10,15,34,34,14,2,1,116


На основе построенной корреляционной таблицы можно сделать следующие выводы:
- Сохранность данных: Суммы маргинальных частот по строкам $n_x$ и столбцам $n_y$ совпадают с результатами одномерной обработки данных в практических работах №1 и №2, а общая сумма частот равна объему выборки $N = 116$, что подтверждает корректность группировки данных в двумерный интервальный ряд.
- Характер связи: Основная масса частот сосредоточена вдоль главной диагонали таблицы. Это визуально подтверждает наличие положительной корреляционной зависимости.
- Теснота связи: Отсутствие значимых частот в ячейках, удаленных от диагонали, свидетельствует о высокой степени согласованности признаков. Это дополняет вывод из задания №1 об однородности выборки.

### 4. Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.

In [14]:
mid_nu, counts_nu, h_nu_val, _ = get_interval_stats(df['nu'])
table1_nu, C_nu, h_nu_final = create_table_1(mid_nu, counts_nu, h_nu_val, "признака nu")
x_nu, d_nu, sigma_nu = compute_main_stats(mid_nu, counts_nu, calculate_moments(table1_nu, h_nu_final), C_nu, h_nu_final, "nu")

mid_e, counts_e, h_e_val, _ = get_interval_stats(df['E'])
table1_e, C_e, h_e_final = create_table_1(mid_e, counts_e, h_e_val, "E")
x_e, d_e, sigma_e = compute_main_stats(mid_e, counts_e, calculate_moments(table1_e, h_e_final), C_e, h_e_final, "E")

n_ij = correlation_table.iloc[:-1, :-1].values

# Условные варианты
u_x = table1_nu[r'$u_i$'].iloc[:-1].values
v_y = table1_e[r'$u_i$'].iloc[:-1].values

N = len(df)

sum_n_x_y = 0
sum_n_u_v = 0
for i in range(len(mid_nu)):
    for j in range(len(mid_e)):
        sum_n_x_y += n_ij[i, j] * mid_nu[i] * mid_e[j]
        sum_n_u_v += n_ij[i, j] * u_x[i] * v_y[j]

# Стандартная формула
r_standard = (sum_n_x_y - N * x_nu * x_e) / (N * sigma_nu * sigma_e)

# Метод условных вариант
u_avg = np.sum(table1_nu[r'$n_i \cdot u_i$'].iloc[:-1]) / N
v_avg = np.sum(table1_e[r'$n_i \cdot u_i$'].iloc[:-1]) / N
sigma_u = np.sqrt(np.sum(table1_nu[r'$n_i \cdot u_i^2$'].iloc[:-1]) / N - u_avg**2)
sigma_v = np.sqrt(np.sum(table1_e[r'$n_i \cdot u_i^2$'].iloc[:-1]) / N - v_avg**2)

r_conditional = (sum_n_u_v - N * u_avg * v_avg) / (N * sigma_u * sigma_v)

# Вывод результатов
display(Markdown("### Сравнение методов расчета коэффициента корреляции"))
comparison_r = pd.DataFrame({
    "Метод": ["Стандартная формула", "Условные варианты"],
    "Значение r_xy": [r_standard, r_conditional]
})
display(comparison_r)
display(Markdown(f"**Разница между методами:** {abs(r_standard - r_conditional):.2e}"))

--- Контроль вычислений для признака nu ---


$\sum n_i (u_i+1)^4 = 3722.0$

$\sum n_i \cdot u_i^4 + 4 \sum n_i \cdot u_i^3 + 6 \sum n_i \cdot u_i^2 + 4 \sum n_i \cdot u_i + N = 3722.0$

Контроль пройден: True


#### Результаты для nu:

,Показатель,Стандартная формула,Условные варианты,Разность
0,Среднее значение,455.8276,455.8276,0.0000
1,Выборочная дисперсия,"3,229.3659","3,229.3659",0.0000


Выборочное СКО ($\sigma_в$): **56.8275**

--- Контроль вычислений для E ---


$\sum n_i (u_i+1)^4 = 2955.0$

$\sum n_i \cdot u_i^4 + 4 \sum n_i \cdot u_i^3 + 6 \sum n_i \cdot u_i^2 + 4 \sum n_i \cdot u_i + N = 2955.0$

Контроль пройден: True


#### Результаты для E:

,Показатель,Стандартная формула,Условные варианты,Разность
0,Среднее значение,128.1636,128.1636,0.0000
1,Выборочная дисперсия,480.7445,480.7445,0.0000


Выборочное СКО ($\sigma_в$): **21.9259**

### Сравнение методов расчета коэффициента корреляции

,Метод,Значение r_xy
0,Стандартная формула,0.8341
1,Условные варианты,0.8341


**Разница между методами:** 4.00e-15

На основе результатов вычислений выборочного коэффициента корреляции можно сделать следующие выводы:
- Корректность расчетов: Значения коэффициента корреляции, полученные стандартным методом и методом условных вариант, практически полностью совпадают. Это подтверждает корректность проведенной обработки данных и эквивалентность использованных подходов.
- Теснота и направление связи: Полученное значение $r_{xy} \approx 0.8341$ свидетельствует о наличии сильной положительной линейной связи между объемным весом $\nu$ и модулем упругости $E$ древесины резонансной ели.
- Интерпретация: Высокое значение коэффициента корреляции подтверждает визуальные наблюдения из задания №3: изменение одного признака в значительной степени обуславливает изменение математического ожидания другого признака, что характерно для корреляционной зависимости.

### 5. Построить доверительный интервал для коэффициента корреляции при уровне значимости $\gamma \in \{0.95,0.99\}$, сделать выводы.

In [15]:
def calculate_correlation_interval(r, n, gamma):
    # Преобразование Фишера
    z_avg = 0.5 * np.log((1 + r) / (1 - r))
    
    # СКО величины z
    sigma_z = 1 / np.sqrt(n - 3)
    
    # Нахождение квантиля lambda (из функции Лапласа)
    lmbd = stats.norm.ppf(0.5 + gamma / 2)
    
    # Доверительный интервал для z
    z_low = z_avg - lmbd * sigma_z
    z_high = z_avg + lmbd * sigma_z
    
    # Обратное преобразование Фишера (переход от z к r)
    r_low = np.tanh(z_low)
    r_high = np.tanh(z_high)
    
    return r_low, r_high, lmbd

r_val = r_standard
n_val = N

results_intervals = []
for g in [0.95, 0.99]:
    low, high, l_val = calculate_correlation_interval(r_val, n_val, g)
    results_intervals.append({
        "Надежность (gamma)": g,
        "lambda": l_val,
        "Нижняя граница": low,
        "Верхняя граница": high
    })

# Вывод результатов
display(Markdown("### Доверительные интервалы для коэффициента корреляции"))
intervals_df = pd.DataFrame(results_intervals)
display(intervals_df)

### Доверительные интервалы для коэффициента корреляции

,Надежность (gamma),lambda,Нижняя граница,Верхняя граница
0,0.9500,1.9600,0.7686,0.8822
1,0.9900,2.5758,0.7439,0.8944


На основе результатов расчета можно сделать следующие выводы:
- Надежность оценок: Построенные интервалы позволяют утверждать, что истинное значение коэффициента корреляции в генеральной совокупности находится в диапазоне выше 0.74.
- Связь точности и надежности: С увеличением надежности $\gamma$ с 0.95 до 0.99 ширина доверительного интервала закономерно увеличивается (с 0.1136 до 0.1505), что ведет к снижению точности оценки при одновременном повышении её достоверности.
- Статистическая значимость: Оба полученных интервала ($\[0,7686;0,8822\] и \[0,7439;0,8944\]$) полностью лежат в положительной области и не включают в себя ноль. Это свидетельствует о высокой вероятности наличия существенной линейной связи между плотностью $\nu$ и модулем упругости $E$.

### 6. Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости $\alpha = 0.05$, сделать выводы.

In [16]:
# Уровень значимости и число степеней свободы
alpha = 0.05
k_df = N - 2

# Вычисление T-статистики
t_obs = (r_standard * np.sqrt(k_df)) / np.sqrt(1 - r_standard**2)

# Критическое значение
t_crit = stats.t.ppf(1 - alpha / 2, k_df)

display(Markdown("### Проверка гипотезы о значимости коэффициента корреляции"))
decision = "отклоняется (связь значима)" if abs(t_obs) > t_crit else "принимается (связь не значима)"

test_results = pd.DataFrame({
    "Показатель": ["Уровень значимости (alpha)", "Степени свободы (k)", "T_набл", "t_крит", "Решение по H0"],
    "Значение": [alpha, k_df, t_obs, t_crit, decision]
})

display(test_results)

if abs(t_obs) > t_crit:
    print(f"Вывод: Так как |{t_obs:.4f}| > {t_crit:.4f}, нулевая гипотеза отвергается.")
    print("Выборочный коэффициент корреляции статистически значим.")
else:
    print(f"Вывод: Так как |{t_obs:.4f}| <= {t_crit:.4f}, нет оснований отвергать нулевую гипотезу.")

### Проверка гипотезы о значимости коэффициента корреляции

,Показатель,Значение
0,Уровень значимости (alpha),0.0500
1,Степени свободы (k),114
2,T_набл,16.1436
3,t_крит,1.9810
4,Решение по H0,отклоняется (связь значима)


Вывод: Так как |16.1436| > 1.9810, нулевая гипотеза отвергается.
Выборочный коэффициент корреляции статистически значим.


На основе полученных результатов можно сделать следующие выводы:
- Значимость связи: отклонение нулевой гипотезы означает, что выборочный коэффициент корреляции статистически значим.
- Обоснованность выводов: полученные данные позволяют утверждать, что выявленная в ходе работы сильная прямая связь между плотностью $\nu$ и модулем упругости $E$ древесины резонансной ели не является следствием случайного подбора данных, а отражает объективную закономерность.
- Применимость модели: значимость коэффициента корреляции служит основой для дальнейшего построения уравнений регрессии и интерпретации экспериментальных наблюдений.

## Выводы

В ходе выполнения практической работы были освоены и реализованы методы корреляционного анализа двумерной выборки экспериментальных данных. Процесс перехода от построения двумерного интервального ряда к расчету числовых характеристик связи позволил подтвердить наличие сильной прямой линейной зависимости между физическими свойствами древесины — объемным весом $\nu$ и модулем упругости $E$.

Применение метода условных вариант продемонстрировало высокую точность и совпадение результатов со стандартным методом расчета, что подтверждает корректность полученных оценок. Использование преобразования Фишера позволило компенсировать асимметрию распределения выборочного коэффициента корреляции и построить доверительные интервалы для генеральной совокупности.

Финальная проверка статистических гипотез с использованием критерия Стьюдента подтвердила, что выявленная взаимосвязь является статистически значимой и не обусловлена случайными факторами выборки.